programs/bpf_loader/src/lib.rs

```diff
@@ -575,89 +670,122 @@ fn process_loader_upgradeable_instruction(
-                programdata_address,
-            } = program.state()?
-            {
-                if programdata_address != *programdata.unsigned_key() {
-                    ic_logger_msg!(log_collector, "Program and ProgramData account mismatch");
+            let new_program_id = {
+                let program = instruction_context.try_borrow_instruction_account(
+                    transaction_context,
+                    upgradeable_ins_acc_idx::Upgrade::Program as usize,
+                )?;
+                if !program.is_executable() {
+                    ic_logger_msg!(log_collector, "Program account not executable");
+                    return Err(InstructionError::AccountNotExecutable);
+                }
+                if !program.is_writable() {
+                    ic_logger_msg!(log_collector, "Program account not writeable");
                     return Err(InstructionError::InvalidArgument);
                 }
-            } else {
-                ic_logger_msg!(log_collector, "Invalid Program account");
-                return Err(InstructionError::InvalidAccountData);
-            }
-
-            let new_program_id = *program.unsigned_key();
+                if program.get_owner() != program_id {
+                    ic_logger_msg!(log_collector, "Program account not owned by loader");
+                    return Err(InstructionError::IncorrectProgramId);
+                }
+                if let UpgradeableLoaderState::Program {
+                    programdata_address,
+                } = program.get_state()?
+                {
+                    if current_programdata_address != programdata_address {
+                        ic_logger_msg!(log_collector, "Program and ProgramData account mismatch");
+                        return Err(InstructionError::InvalidArgument);
+                    }
+                } else {
+                    ic_logger_msg!(log_collector, "Invalid Program account");
+                    return Err(InstructionError::InvalidAccountData);
+                }
+                *program.get_key()
+            };
 
             // Verify Buffer account
-
-            if let UpgradeableLoaderState::Buffer { authority_address } = buffer.state()? {
-                if authority_address != Some(*authority.unsigned_key()) {
+            let buffer = instruction_context.try_borrow_instruction_account(
+                transaction_context,
+                upgradeable_ins_acc_idx::Upgrade::Buffer as usize,
+            )?;
+            if let UpgradeableLoaderState::Buffer { authority_address } = buffer.get_state()? {
+                if authority_address != current_upgrade_authority_address {
                     ic_logger_msg!(log_collector, "Buffer and upgrade authority don't match");
                     return Err(InstructionError::IncorrectAuthority);
                 }
-                if authority.signer_key().is_none() {
+                if !instruction_context.is_signer(
```

Comment:
The difference in index base here can lead to confusion

Reply:
True and already on my TODO list of things to iron out once I refactored all builtin programs.
Did not want to bring that change into this PR as well.